### PR TITLE
common/options: convert a millisecs opt to a chrono::milliseconds and cleanups

### PR DESCRIPTION
**Pull request by tchaikov**
Fixes: https://tracker.ceph.com/issues/51375

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by tchaikov** [comment]
jenkins test make check

**Comment by tchaikov** [comment]
jenkins test api

**Comment by tchaikov** [comment]
jenkins test api

**Comment by tchaikov** [comment]
jenkins test api

**Comment by tchaikov** [comment]
jenkins test api

**Comment by tchaikov** [comment]
jenkins test api

**Review by batrick — APPROVED** [review]
\o/

**Comment by tchaikov** [comment]
- https://pulpito.ceph.com/kchai-2021-06-27_13:33:07-rados-wip-kefu-testing-2021-06-27-1907-distro-basic-smithi/
- https://pulpito.ceph.com/kchai-2021-06-29_07:34:48-rados-wip-kefu-testing-2021-06-27-1907-distro-basic-smithi/

failures either tracked or fixed by


- https://tracker.ceph.com/issues/5024
- https://tracker.ceph.com/issues/50151
- #42084

File: src/common/config.cc [hunk 1]
```diff
@@ -202,7 +202,7 @@ md_config_t::md_config_t(ConfigValues& values,
       }
       // We call pre_validate as a sanity check, but also to get any
       // side effect (value modification) from the validator.
-      std::string *def_str = boost::get<std::string>(&default_val);
+      auto def_str = std::get_if<std::string>(&default_val);
```
Comment (rzarzynski) [inline src/common/config.cc:203]: nit: would `auto*` improve readability here?

File: src/common/config.h [hunk 2]
```diff
@@ -70,7 +70,7 @@ extern const char *ceph_conf_level_name(int level);
  */
 struct md_config_t {
 public:
-  typedef boost::variant<int64_t ConfigValues::*,
+  typedef std::variant<int64_t ConfigValues::*,
                          uint64_t ConfigValues::*,
```
Comment (rzarzynski) [inline src/common/config.h:74]: nit: indentation.

File: src/common/config.h [hunk 3]
```diff
@@ -361,7 +361,7 @@ const T md_config_t::get_val(const ConfigValues& values,
   return std::get<T>(this->get_val_generic(values, key));
 }
 
-inline std::ostream& operator<<(std::ostream& o, const boost::blank& ) {
+inline std::ostream& operator<<(std::ostream& o, const std::monostate& ) {
```
Comment (rzarzynski) [inline src/common/config.h:364]: nit: not know but how about killing the extra space before `)`?

File: src/common/config.cc [hunk 4]
```diff
@@ -1474,34 +1472,33 @@ struct get_size_visitor : public boost::static_visitor<Size>
...
   public:
 
-  assign_visitor(Config *conf_, Option::value_t val_)
+  assign_visitor(ConfigValues *conf_, Option::value_t val_)
     : conf(conf_), val(val_)
   {}
 
   template <typename T>
-  void operator()(T Config::* ptr) const
+  void operator()(T ConfigValues::* ptr) const
```
Comment (rzarzynski) [inline src/common/config.cc:1486]: For the posterity: it's a templated pointer-to-member.

**Review by rzarzynski — APPROVED** [review]
LGTM!

> If `which() == rhs.which()` then: `content_this < content_rhs` (...). Otherwise: `which() < rhs.which()`.

Well, this is a very good illustration on how to surprise a developer.

File: src/common/config.cc (hunk 1 above)
Comment (tchaikov) [reply src/common/config.cc:203]: updated.

File: src/common/config.h (hunk 2 above)
Comment (tchaikov) [reply src/common/config.h:74]: fixed.

File: src/common/config.h (hunk 3 above)
Comment (tchaikov) [reply src/common/config.h:364]: removed.

**Comment by tchaikov** [comment]
@rzarzynski fixed and repushed.

**Review by rzarzynski — APPROVED** [review]

**Comment by tchaikov** [comment]
jenkins test make check

**Pull request merged**